### PR TITLE
Support multiple accounts

### LINKED ISSUE #181: Multiple "firebase login" per client account
Apologies if this was already raised. It would be nice for `firebase login` to not have to be client account wide (i.e. `~/.config/configstore/firebase-tools.json`). I use an OSX machine, and deploy to multiple firebase accounts. Switching users on OSX is a PITA (and disrupts communication apps). Logging in and out of firebase is also not ideal. Ideally, the config would be stored in the project root (and I'd just `.gitignore` it).

Let me know where to look at a high level and I'll submit a PR. Or perhaps this already exists and I'm just using it wrong.

### LINKED ISSUE #1906: Feature Request: Method for checking which user is logged in – without prompts
Hi!

This is a technical feature request for the CLI.

I use Firebase with two different accounts, and I frequently switch between them. This means a lot of `firebase logout` and `firebase login`, and that's fine. But it is easy to forget this before I do a deploy. And when I forget this, I might have waited multiple minutes for my project to build, only for `firebase deploy` to fail because I'm logged in with the wrong user.

What I would like is something like `firebase login --check` which either 1) prints `Not logged in` if no user is logged in, or, 2) the email address of the logged in user. It should exit with a code > 0 if not logged in, 0 if logged in.

This would make me able to create a step in my build pipeline that checks that I'm logged in with the right user, and trigger an early exit if the wrong user is logged in. I could even extend it by creating a bash script that matches the logged in email with my company's domain, use regex, and so on.

The existing `firebase login` command does print the email of the logged-in user, but it will prompt the user and perform a login if not logged in. All in all, this is not very suitable for parsing/scripting.

Thanks for listening!

PS: I know GitHub is not the place for feature requests, as clearly stated in the Issue Template. I did submit this through the feature request form, but the response I got was that I should post this here instead... So if it's correct that feature requests for the CLI should be posted here, I'd suggest clarifying this in the issue template.

### PR DESCRIPTION
**Pull request by samtstern**
### Description

Implement multiple authentication (see [Internal API proposal](https://docs.google.com/document/d/1SsRDj79mVu-Su7EBpFdKf-SAYwwq2nszHKjJB5ORuU0/edit?usp=sharing&resourcekey=0-535ICGcabfQa7SguXhPIPA))

Fixes #1906
Fixes #181 

### Scenarios Tested

**login:add**

```
$ firebase login:add

Visit this URL on this device to log in:
https://accounts.google.com/o/oauth2/auth?client_id=563584335869-fgrhgmd47bqnekij5i8b5pr03ho849e6.apps.googleusercontent.com&scope=email%20openid%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloudplatformprojects.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Ffirebase%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform&response_type=code&state=75254219&redirect_uri=http%3A%2F%2Flocalhost%3A9005

Waiting for authentication...

✔  Success! Added account samsternota@gmail.com
```

**login:list**
```
$ firebase login:list
Logged in as samsterntest@gmail.com

Other available accounts (switch with login:use)
 - samstern@google.com
 - samsternota@gmail.com
```

**login:use**
```
$ firebase login:use samsternota@gmail.com
✔  Set default account samsternota@gmail.com for current project directory.

# ...

$ firebase login:list
Logged in as samsternota@gmail.com

Other available accounts (switch with login:use)
 - samstern@google.com
 - samsterntest@gmail.com
```

**logout**
```
$ firebase logout
✔  Logged out from samstern@google.com
✔  Logged out from samsterntest@gmail.com
✔  Logged out from samsternota@gmail.com
```

**logout (primary account**
```
$ firebase logout samstern@google.com
? You are logging out of your default account, which account should become the 
new default? samsternota@gmail.com
✔  Logged out from samstern@google.com
✔  Setting default account to "samsternota@gmail.com"

# ...

$ firebase login:list
Logged in as samsternota@gmail.com

Other available accounts (switch with "firebase login:use")
 - samsterntest@gmail.com
```

**init**
```
$ firebase init database

     ######## #### ########  ######## ########     ###     ######  ########
     ##        ##  ##     ## ##       ##     ##  ##   ##  ##       ##
     ######    ##  ########  ######   ########  #########  ######  ######
     ##        ##  ##    ##  ##       ##     ## ##     ##       ## ##
     ##       #### ##     ## ######## ########  ##     ##  ######  ########

You're about to initialize a Firebase project in this directory:

  /private/var/folders/xl/6lkrzp7j07581mw8_4dlt3b000643s/T/tmp.2OgG9jvU

Before we get started, keep in mind:

  * You are currently outside your home directory


=== Account Setup

Which account do you want to use for this project? Choose an account or add a new one now

? Please select an option: samsternota@gmail.com

✔  Using account: samsternota@gmail.com

=== Project Setup

First, let's associate this project directory with a Firebase project.
You can create multiple project aliases by running firebase use --add, 
but for now we'll just set up a default project.

? Please select an option: Use an existing project
? Select a default Firebase project for this directory: fir-dumpster-ota (fir-du
mpster-ota)
i  Using project fir-dumpster-ota (fir-dumpster-ota)

=== Database Setup
i  database: ensuring required API firebasedatabase.googleapis.com is enabled...
⚠  database: missing required API firebasedatabase.googleapis.com. Enabling now...
✔  database: required API firebasedatabase.googleapis.com is enabled

? It seems like you haven’t initialized Realtime Database in your project yet. D
o you want to set it up? No

Firebase Realtime Database Security Rules allow you to define how your data should be
structured and when your data can be read from and written to.

? What file should be used for Realtime Database Security Rules? database.rules.
json
✔  Default rules have been written to database.rules.json.
Future modifications to database.rules.json will update Realtime Database Security Rules when you run
firebase deploy.

i  Writing configuration info to firebase.json...
i  Writing project information to .firebaserc...
i  Writing gitignore file to .gitignore...

✔  Firebase initialization complete!
```

**--account flag (denied)**
```
$ firebase --account=samstern@google.com --project=fir-dumpster-ota functions:config:get

Error: HTTP Error: 403, The caller does not have permission
```

**--account flag (succeed)**
```
$ firebase --account=samsternota@gmail.com --project=fir-dumpster-ota functions:config:get
{
  "foo": {
    "baz": "qux"
  }
}
```

### Sample Commands

[Internal API proposal](https://docs.google.com/document/d/1SsRDj79mVu-Su7EBpFdKf-SAYwwq2nszHKjJB5ORuU0/edit?usp=sharing&resourcekey=0-535ICGcabfQa7SguXhPIPA)

### CODE REVIEW

File: src/commands/logout.js [hunk 1]
```diff
@@ -1,50 +0,0 @@
-"use strict";
```
Comment (samtstern) [inline src/commands/logout.js:1]: Re-written in TS

**Comment by joehan** [comment]
Did some testing, and noticed that the `--account` flag doesn't work like the `--project` flag does for `firebase init`:
Screen Shot 2021-03-09 at 9 13 37 AM
Based on the API proposal, I think this should not prompt me for which account to use.

File: src/commands/login-use.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import * as _ from "lodash";
```
Comment (joehan) [inline src/commands/login-use.ts:1]: Unused import

File: src/commands/login-use.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import * as clc from "cli-color";
```
Comment (joehan) [inline src/commands/login-use.ts:2]: Unused import?

File: src/commands/logout.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
+"use strict";
```
Comment (joehan) [inline src/commands/logout.ts:1]: Can get rid of this since this is ts now.

File: src/commands/logout.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+import { Command } from "../command";
+import * as logger from "../logger";
+import * as clc from "cli-color";
+
+import * as utils from "../utils";
+import * as api from "../api";
+import * as auth from "../auth";
+import * as _ from "lodash";
```
Comment (joehan) [inline src/commands/logout.ts:10]: unused import

File: src/init/features/account.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+/**
+ * Sets up the project default account.
+ * @param setup A helper object to use for the rest of the init features.
+ * @param config Configuration for the project.
+ * @param options Command line options.
+ */
+export async function doSetup(setup: any, config: any, options: any): Promise<void> {
+  let account: Account | undefined = undefined;
```
Comment (joehan) [inline src/init/features/account.ts:62]: Nit:
```suggestion
  let account: Account | undefined;
```

**Review by joehan — CHANGES_REQUESTED** [review]
Some cleanup, but most the core logic seems good to me

File: src/auth.ts [hunk 7]
```diff
@@ -394,10 +631,12 @@ async function refreshTokens(
       res.body
     );
 
-    const currentRefreshToken = configstore.get("tokens")?.refresh_token;
-    if (refreshToken === currentRefreshToken) {
-      configstore.set("tokens", lastAccessToken);
+    const account = findAccountByRefreshToken(refreshToken);
+    if (account && lastAccessToken) {
+      account.tokens = lastAccessToken;
```
[the remark this answers is not in the document]
Comment (samtstern) [reply src/auth.ts:636]: Trying to keep terminology similar where I can. The `"tokens"` key has been used in configstore to refer to the object of type `Tokens` which contains access and refresh tokens. This is the same type.

File: src/commands/login-list.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    const otherAccounts = allAccounts.filter((a) => a.user.email !== user.email);
+    if (otherAccounts.length > 0) {
+      logger.info();
+      logger.info("Other available accounts (switch with login:use)");
+      for (const a of otherAccounts) {
+        logger.info(` - ${a.user.email}`);
+      }
+    }
+
+    return auth;
```
[the remark this answers is not in the document]
Comment (samtstern) [reply src/commands/login-list.ts:33]: Truly have no idea, was just copying `login` ... I'll remove it.

File: src/auth.ts [hunk 9]
```diff
@@ -348,15 +541,59 @@ function haveValidTokens(refreshToken: string, authScopes: string[]) {
...
   }
+
+  // Check all additional users
+  const additionalAccounts = getAdditionalAccounts();
+  const remainingAccounts = additionalAccounts.filter((a) => a.user.email !== account.user.email);
+  configstore.set("additionalAccounts", remainingAccounts);
+
+  // Clear any matching project defaults
+  const activeAccounts = configstore.get("activeAccounts") || {};
+  for (const projectDir of Object.keys(activeAccounts)) {
```
[the remark this answers is not in the document]
Comment (samtstern) [reply src/auth.ts:561]: I changed it but then eslint yelled at me about guarding it, which I think would be uglier:
https://eslint.org/docs/rules/guard-for-in

[the remark this answers is not in the document]
Comment (joehan) [reply src/auth.ts:561]: Ahhh, yeah, I ran into that same rule recently. Maybe we should get rid of the rule, but I agree this is less ugly than adding an eslint-ignore or guarding it

File: src/commands/index.js [hunk 10]
```diff
@@ -107,7 +107,10 @@ module.exports = function (client) {
   client.hosting.sites.list = loadCommand("hosting-sites-list");
   client.init = loadCommand("init");
   client.login = loadCommand("login");
+  client.login.add = loadCommand("login-add");
   client.login.ci = loadCommand("login-ci");
+  client.login.list = loadCommand("login-list");
+  client.login.list = loadCommand("login-use");
```
Comment (joehan) [inline src/commands/index.js:113]:
```suggestion
  client.login.use = loadCommand("login-use");
```

File: src/defaultCredentials.ts [hunk 11]
```diff
@@ -98,12 +97,11 @@ function credFilePath(): string | undefined {
...
 
-  return path.join(fbtConfigDir, `${userEmailSlug()}_application_default_credentials.json`);
+  return path.join(fbtConfigDir, `${userEmailSlug(user)}_application_default_credentials.json`);
 }
 
-function userEmailSlug(): string {
-  const user = configstore.get("user");
-  const email = user && user.email ? user.email : "unknown_user";
+function userEmailSlug(user: User): string {
+  const email = user.email ? user.email : "unknown_user";
```
Comment (joehan) [inline src/defaultCredentials.ts:104]:
```suggestion
  const email = user.email || "unknown_user";
```

File: src/management/projects.ts [hunk 12]
```diff
@@ -147,6 +147,7 @@ async function selectProjectInteractively(
   }
   if (nextPageToken) {
     // Prompt user for project ID if we can't list all projects in 1 page
+    logger.debug(`Found ${projects.length} projects, selecting via prompt`);
```
Comment (joehan) [inline src/management/projects.ts:150]:
```suggestion
    logger.debug(`Found more than ${projects.length} projects, selecting via prompt`);
```

**Review by joehan — APPROVED** [review]
LGTM once you fix the typo in commands/index.js

Probably deserves a second set of eyes though given the scope.

File: CHANGELOG.md [hunk 13]
```diff
@@ -0,0 +1 @@
+- Adds support for multiple accounts via new commands `login:use`, `login:add` and `login:list`.
```
Comment (bkendall) [inline CHANGELOG.md:1]: This feels like a feature that could do with (a) documentation in the README and (b) documentation in FireSite

File: scripts/emulator-tests/functionsEmulator.spec.ts [hunk 14]
```diff
@@ -25,6 +25,7 @@ if ((process.env.DEBUG || "").toLowerCase().indexOf("spec") >= 0) {
 
 const functionsEmulator = new FunctionsEmulator({
   projectId: "fake-project-id",
+  account: undefined,
```
Comment (bkendall) [inline scripts/emulator-tests/functionsEmulator.spec.ts:28]: Why are we compelled to add this? `undefined` should be the default if it's not set. If we need to specify this, can we add a comment as to why? (same in the file below)

File: src/auth.ts [hunk 15]
```diff
@@ -54,6 +76,173 @@ interface GitHubAuthResponse {
...
+    user,
+    tokens,
+  };
+}
+
+/**
+ * Get the default account associated with a project directory, or the global default.
+ * @param projectDir the Firebase project directory.
+ */
+export function getProjectDefaultAccount(projectDir?: string | null): Account | undefined {
```
Comment (bkendall) [inline src/auth.ts:103]: Can `projectDir` really be `null`? What does that mean rather than just an empty string?

File: src/auth.ts [hunk 16]
```diff
@@ -54,6 +76,173 @@ interface GitHubAuthResponse {
...
+  }
+
+  const allAccounts = getAllAccounts();
+  return allAccounts.find((a) => a.user.email === email);
+}
+
+/**
+ * Get all authenticated accounts _besides_ the default account.
+ */
+export function getAdditionalAccounts(): Account[] {
```
Comment (bkendall) [inline src/auth.ts:122]: I'm noticing that all these new functions are exported - is that necessary? Some of these seem like helpers...

File: src/auth.ts [hunk 17]
```diff
@@ -54,6 +76,173 @@ interface GitHubAuthResponse {
...
+ * @param useLocalhost should the flow be interactive or code-based?
+ * @param email an optional hint to use for the google account picker
+ */
+export async function loginAdditionalAccount(useLocalhost: boolean, email?: string) {
+  // Log the user in using the passed email as a hint
+  const result = await loginGoogle(useLocalhost, email);
+
+  // The JWT library can technically return a string, even though it never should.
+  if (typeof result.user === "string") {
+    throw new FirebaseError("Failed to parse auth response, see debug log.", { exit: 1 });
```
Comment (bkendall) [inline src/auth.ts:206]: `exit: 1` is the default and doesn't need to be specified.

File: src/auth.ts [hunk 18]
```diff
@@ -348,15 +541,59 @@ function haveValidTokens(refreshToken: string, authScopes: string[]) {
...
   }
+
+  // Check all additional users
+  const additionalAccounts = getAdditionalAccounts();
+  const remainingAccounts = additionalAccounts.filter((a) => a.user.email !== account.user.email);
+  configstore.set("additionalAccounts", remainingAccounts);
+
+  // Clear any matching project defaults
+  const activeAccounts: Record<string, string> = configstore.get("activeAccounts") || {};
+  for (const projectDir of Object.keys(activeAccounts)) {
```
Comment (bkendall) [inline src/auth.ts:561]: You _could_ use `Object.entries` here, if you want to save a line...

File: src/commands/login-add.ts [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+      throw new FirebaseError("Cannot run login:add in non-interactive mode.", { exit: 1 });
+    }
+
+    const account = auth.getGlobalDefaultAccount();
+
+    // "login" asks for the data collection preference, we only want to do that in one place
+    if (!account) {
+      throw new FirebaseError(
+        "No existing accounts found, please run login to add your first account",
+        { exit: 1 }
```
Comment (bkendall) [inline src/commands/login-add.ts:26]: `exit: 1` is the default and doesn't need to be specified (throughout)

File: src/commands/login-list.ts [hunk 20]
```diff
@@ -0,0 +1,30 @@
+import { User } from "../auth";
+import { Command } from "../command";
+import { logger } from "../logger";
+import * as utils from "../utils";
+import * as auth from "../auth";
+
+module.exports = new Command("login:list")
+  .description("List authorized CLI accounts")
```
Comment (bkendall) [inline src/commands/login-list.ts:8]:
```suggestion
  .description("list authorized CLI accounts")
```

File: src/commands/login-list.ts [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+      utils.logWarning(`No authorized accounts, run "firebase login"`);
+      return;
+    }
+
+    logger.info(`Logged in as ${user.email}`);
+
+    const otherAccounts = allAccounts.filter((a) => a.user.email !== user.email);
+    if (otherAccounts.length > 0) {
+      logger.info();
+      logger.info("Other available accounts (switch with login:use)");
```
Comment (bkendall) [inline src/commands/login-list.ts:23]:
```suggestion
      logger.info("Other available accounts (switch with login:use):");
```

I'd also format `login:use` as a command... which style that is off hand I'm not certain

File: src/commands/login-use.ts [hunk 22]
```diff
@@ -0,0 +1,27 @@
+import { Command } from "../command";
+import * as utils from "../utils";
+import * as auth from "../auth";
+import { FirebaseError } from "../error";
+
+module.exports = new Command("login:use <email>")
+  .description("Set the default account to use for this project directory")
```
Comment (bkendall) [inline src/commands/login-use.ts:7]:
```suggestion
  .description("set the default account to use for this project directory")
```

File: src/commands/login-use.ts [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+    const allAccounts = auth.getAllAccounts();
+    const accountExists = allAccounts.some((a) => a.user.email === email);
+    if (!accountExists) {
+      throw new FirebaseError(
+        `Account ${email} does not exist, run login:list to see valid accounts`,
+        { exit: 1 }
+      );
+    }
+
+    const projectDir = options.projectRoot as string | null;
```
Comment (bkendall) [inline src/commands/login-use.ts:18]: Why not just `options.projectRoot as string || ""`? (I'm not a fan of using `null` if I can help it, as you may have guessed by now)

File: src/test/multiAuth.spec.ts [hunk 24]
```diff
@@ -0,0 +1,144 @@
+import { expect } from "chai";
+import * as sinon from "sinon";
+
+import * as auth from "../auth";
+import { configstore } from "../configstore";
+
+describe("multiAuth", () => {
```
Comment (bkendall) [inline src/test/multiAuth.spec.ts:7]: The name of this file (and this block) are confusing me. The `.spec` files should match source files in the repo. Could this match something better? `auth.spec.ts`?

**Review by bkendall — CHANGES_REQUESTED** [review]

File: scripts/emulator-tests/functionsEmulator.spec.ts (hunk 14 above)
Comment (samtstern) [reply scripts/emulator-tests/functionsEmulator.spec.ts:28]: I was mostly using this to have `tsc` help me find all the places in the codebase where these args were passed, you're right that I can make it optional now.

File: src/auth.ts (hunk 15 above)
Comment (samtstern) [reply src/auth.ts:103]: This comes from `detectProjectRoot` which returns `string | null` ... it's a nasty type but I want the code to be honest about what `options.projectRoot` can be and both `undefined` and `null` are possible.

File: src/auth.ts (hunk 16 above)
Comment (samtstern) [reply src/auth.ts:122]: They're all used in at least one other file.

File: CHANGELOG.md (hunk 13 above)
Comment (samtstern) [reply CHANGELOG.md:1]: cc @rachelsaunders I added a README section, can you take a look?

File: src/commands/login-use.ts (hunk 23 above)
Comment (samtstern) [reply src/commands/login-use.ts:18]: This is the actual type based on `detectProjectRoot`, the `as` is just because I don't want to use `any`.

**Review by bkendall — APPROVED** [review]
I didn't hammer every corner, I'm sure, but I pulled this down and played with it a bit, init-ing and deploying things. Looks good to me!

**Pull request merged**